### PR TITLE
Fabric Interop Node SDK Support for HTLC-based Asset Exchanges

### PR DESCRIPTION
**Pull request by VRamakrishna**

### CODE REVIEW

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (VRamakrishna) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:127]: Yes, we can allow the user to submit a preimage, and I can add a parameter for that. In case a user doesn't supply it, I can use a random generation function (though I know you prefer throwing an error.)

But are there any cons to the approach that is implemented in the above code? The value of the preimage is limited to its ability to validate a claim by either party, no? So does it matter who generates it and where? It seems to me providing this functionality in code reduces one more thing the user needs to do. After all, the user (or an agent in the app) is simply interested in effecting an asset exchange with integrity, and the preimage is as important here (I think) as TLS keys would be in a typical secure web connection (where the endpoints should only care about the data getting exchanged and not the mechanism by which channel confidentiality is maintained.)

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (VRamakrishna) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:207]: Yes, this bothered me while I was coding the base class too. That's why I added the functions to declare the total count of fungible assets that are currently lockable. I agree we need some "contract id" that uniquely identifies a particular instance. Let's discuss that on Monday? It'll impact all the modules but should be easy to add to the API.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (VRamakrishna) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:109]: Yes, I thought logs were sufficient to track errors. The Fabric gateway API works such that it either returns a chaincode execution result or throws an error if something goes wrong in the network. So that error will automatically propagate up to the app. And if you see the invocation command, we catch and throw that error already. Since these functions are executed within the application's process, throwing a log error here or in the calling function doesn't make a difference IMO.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (ermyas) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:155]: I think this loss in precision could be a problem, and one I suspect we would need to address at some point (hopefully without breaking compatibility). There is an open PR to address this issue by enabling protobuf to use `BigInt` instead of `number` for 64bit ints https://github.com/protobufjs/protobuf.js/pull/1557
We can defer this for now.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (ermyas) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:129]: I don't think time in the context of HTLCs is an implementation detail that the user can be oblivious to. While it would be ideal from a UX perspective if the user did not have to care about such parameters, this specific parameter dictates the behaviour of the user and is one they have to be conscious of (i.e. they have to ensure to manually perform certain actions within the bounds of this parameter or risk forfeiting the exchange). For this reason, I think the value of this parameter should be a deliberate choice by the user rather than an arbitrary default that they might not know about. In future, as we automate part of the currently manual steps of the HTLC flow, we can consider making these details transparent to the user

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (ermyas) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:109]: Not sure I quite got this, but I'll defer to your judgment.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (ermyas) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:127]: Right, I am not opposed to offering a convenient way for a user to generate a preimage. As noted above, I think that would be useful to do. However, given this is an important parameter, I think we want to make sure that we minimize our assumptions about the user's intents. Inferring that a user wants an automated preimage created if the parameter was empty feels a bit risky. Such a scenario could be a result of a trivial bug in the calling context and not an explicit intent by the user. Hence, why I'd recommend that we either make the pre-image parameter mandatory and thus force the calling context to first generate a pre-image using our generator (or a user-supplied one) before calling this API, or we offer an alternate function that makes clear that a generated pre-image will be used. These approaches can ensure that the calling context is clear about its intentions.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (ermyas) [inline sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:122]: should we be using a cryptographically secure pseudo-random number generator as opposed to `Math.random()`?

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (VRamakrishna) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:127]: The function is already returning a generated preimage to the user, so I don't think there will be any ambiguity in the calling program about what preimage was used to finally generate the hash. I don't think we need a separate function. If the return value is empty and there is no error, the caller knows its the supplied preimage that was used.

In fact, the Fabric CA client-server registration protocol follows this pattern. When a registrar tries to enroll a new user, it gets back a response with a user ID and an auto-generated secret (password). From a security perspective, as long as the channel is confidential, I think this passes muster. In any case, it's going to be a simple function call in our scenario.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (VRamakrishna) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts]: Done.

File: sdks/fabric/interoperation-node-sdk/src/AssetManager.ts
Comment (ermyas) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts:137]: While I agree that time is an even more sensitive parameter in protocols that have long confirmation times and probabilistic finality, it is still an important one even in permissioned networks, where we might take shorter confirmation times and instant finality for granted. The reason is that a good portion of what that time specifies is the pace at which a person is expected to action each step of the flow. This is something the party would want to be explicit about. Otherwise, we risk making undue assumptions i.e. that the party is trigger ready to execute each step.

An example of where such an assumption might not hold for instance is when considering the party is a large enterprise, with defined internal approval workflows that involve multiple parties signing off before any action that commits the organization is performed. In a solution, we recently developed, for instance, every transaction that comes to the blockchain required an off-chain approval workflow that adheres to the Four-Eye or Six-Eye principle (i.e. 2 parties or 3 parties in the organization sign-off respectively). If an HTLC agreement involving significant amounts of money is set up, it could be that each step of the HTLC process involves multiple off-chain approvals, before being submitted on-chain. This basically means a significant portion of the HTLC flow operates on slower human time, potentially involving a number of parties. A parameter that dictates the pace of this reality has to be deliberately set and agreed upon.

The other minor thing I would add is that from an engineering perspective if one provides a parameter to a function (albeit an invalid one) and the function silently ignores it and sets its own, it would seem to me that the behavior of the function violates the principle of least surprise :-)

Comment (VRamakrishna) [reply sdks/fabric/interoperation-node-sdk/src/AssetManager.ts]: Done.

File: core/network/fabric-interop-cc/interfaces/asset-mgmt/asset_locks_contract.go
Comment (ermyas) [reply core/network/fabric-interop-cc/interfaces/asset-mgmt/asset_locks_contract.go:72]: note that the two reasons we are not guaranteeing events here are because 1) how we treat a failure of `setEvent` in the code above 2) we can only have a [single event in a transaction in Fabric](https://github.com/hyperledger/fabric/blob/1147977a810ac58c97bd9429b052cc9b56e496b0/docs/source/developapps/transactioncontext.md), so if the application chaincode that utilizes this (that we don't have control over), also emits an event, then ours would be replaced

**Review by sanvenDev — APPROVED** [review]

File: core/network/fabric-interop-cc/contracts/interop/manage_assets.go
Comment (VRamakrishna) [reply core/network/fabric-interop-cc/contracts/interop/manage_assets.go]: Refactored some of the code using your above pattern above. Will fix the rest in a subsequent PR.

**Pull request merged**
